modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java

```diff
@@ -160,31 +162,31 @@ private void ensureOpen() {
     }
 
     private void reopenStreamOrFail(IOException e) throws IOException {
-        if (attempt >= maxAttempts) {
-            logger.debug(
-                () -> format(
-                    "failed reading [%s/%s] at offset [%s], attempt [%s] of [%s], giving up",
-                    blobStore.bucket(),
-                    blobKey,
-                    start + currentOffset,
-                    attempt,
-                    maxAttempts
-                ),
-                e
-            );
-            throw addSuppressedExceptions(e);
+        final int maxAttempts = blobStore.getMaxRetries() + 1;
+
+        final long meaningfulProgressSize = Math.max(1L, blobStore.bufferSizeInBytes() / 100L);
+        final long currentStreamProgress = Math.subtractExact(Math.addExact(start, currentOffset), currentStreamFirstOffset);
+        if (currentStreamProgress >= meaningfulProgressSize) {
+            failuresAfterMeaningfulProgress += 1;
         }
-        logger.debug(
-            () -> format(
-                "failed reading [%s/%s] at offset [%s], attempt [%s] of [%s], retrying",
-                blobStore.bucket(),
-                blobKey,
-                start + currentOffset,
-                attempt,
-                maxAttempts
-            ),
-            e
+        final Supplier<String> messageSupplier = () -> format(
+            """
+                failed reading [%s/%s] at offset [%s]; this was attempt [%s] to read this blob which yielded [%s] bytes; in total \
+                [%s] of the attempts to read this blob have made meaningful progress; the maximum number of attempts is [%s]""",
```

Comment:
This log line will be a bit weird because it will say something like:

> this was attempt 103; in total 100 of the attempts made meaningful progress; the maximum number of attempts is 3.

Feels a bit weird that 103 >> 3. Maybe to make it more clear you could say at the end "the maximum number of attempts **(without meaningful progress)** is 3".

Reply:
:+1: good point